### PR TITLE
Support of Grok1 Model

### PR DESCRIPTION
**Pull request by quic-amitraj**
Support of Grok1 Model

### CODE REVIEW

File: QEfficient/transformers/models/grok_1/modeling_grok1.py [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+        hidden_states: torch.Tensor,
+        layer_idx: int,
+        attention_mask: Optional[torch.Tensor] = None,
+        position_ids: Optional[torch.LongTensor] = None,
+        past_key_value: Optional[Tuple[torch.Tensor]] = None,
+        batch_index: Optional[torch.LongTensor] = None,
+        output_attentions: bool = False,
+        use_cache: bool = False,
+        **kwargs,
+    ) -> Tuple[torch.Tensor, Optional[torch.Tensor], Optional[Tuple[torch.Tensor]]]:
```
Comment (quic-hemagnih) [inline QEfficient/transformers/models/grok_1/modeling_grok1.py:47]: Please add doc strings for all the modules (input, ouput)

File: QEfficient/transformers/models/grok_1/modeling_grok1.py [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+        key_states = repeat_kv(key_states, self.num_key_value_groups)
+        value_states = repeat_kv(value_states, self.num_key_value_groups)
+
+        attn_weights = torch.matmul(query_states, key_states.transpose(2, 3)).to(torch.float)
+        attn_weights = attn_weights * self.attn_output_multiplier
+        attn_weights = self.max_attn_val * F.tanh(attn_weights / self.max_attn_val)
+
+        if attention_mask is not None:
+            attn_weights = torch.where(attention_mask, torch.tensor(-10000.0, dtype=torch.float32), attn_weights)
+
```
Comment (quic-hemagnih) [inline QEfficient/transformers/models/grok_1/modeling_grok1.py:84]: Please refrain using the magic numbers. like -10000.0

Comment (quic-hemagnih) [reply QEfficient/transformers/models/grok_1/modeling_grok1.py:84]: Should we use torch.full_like inspite of torch.tensor to create a tensor here.

File: QEfficient/transformers/models/grok_1/modeling_grok1.py [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+                f" {attn_output.size()}"
+            )
+
+        attn_output = attn_output.transpose(1, 2).contiguous()
+        attn_output = attn_output.reshape(bsz, q_len, -1)
+
+        attn_output = self.o_proj(attn_output)
+
+        if not output_attentions:
+            attn_weights = None
```
Comment (quic-hemagnih) [inline QEfficient/transformers/models/grok_1/modeling_grok1.py:100]: We can simply return from here also

File: QEfficient/transformers/models/grok_1/modeling_grok1.py [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+            torch.nn.functional.one_hot(selected_experts[:, 0], num_classes=self.num_experts).bool().T.unsqueeze(-1)
+        )
+        awesome_experts_mask_2 = (
+            torch.nn.functional.one_hot(selected_experts[:, 1], num_classes=self.num_experts).bool().T.unsqueeze(-1)
+        )
+
+        gateupout1 = torch.zeros(hidden_states.shape[0], 32768)  # T, hs
+        gateupout2 = torch.zeros(hidden_states.shape[0], 32768)  # T, hs
+        for expert_idx in range(self.num_experts):
+            expert_layer = self.experts[expert_idx]
```
[the remark this answers is not in the document]
Comment (quic-hemagnih) [reply QEfficient/transformers/models/grok_1/modeling_grok1.py:124]: def forward(self, hidden_states: torch.Tensor):
    batch_size, sequence_length, hidden_dim = hidden_states.shape
    hidden_states = hidden_states.view(-1, hidden_dim)  # [T, D]
    num_tokens = hidden_states.size(0)

    router_logits = self.gate(hidden_states)  # [T, E]
    routing_weights = F.softmax(router_logits, dim=1, dtype=torch.float)
    topk_weights, topk_indices = torch.topk(routing_weights, self.top_k, dim=-1)  # [T, 2]

    # Prepare output buffer
    expert_outputs = torch.zeros_like(hidden_states)

    for k in range(self.top_k):
        expert_ids = topk_indices[:, k]  # [T]
        weights = topk_weights[:, k]     # [T]

        for expert_idx in range(self.num_experts):
            # Mask for tokens assigned to this expert
            mask = expert_ids == expert_idx
            if not mask.any():
                continue

            selected_inputs = hidden_states[mask]  # [N, D]
            selected_weights = weights[mask].unsqueeze(-1)  # [N, 1]

            expert = self.experts[expert_idx]
            expert_output = expert.act_fn(expert.linear(selected_inputs)) * expert.linear_v(selected_inputs)
            expert_output = expert.linear_1(expert_output)

            # Weighted output
            expert_outputs[mask] += expert_output * selected_weights

    hidden_states = expert_outputs.view(batch_size, sequence_length, hidden_dim)
    return hidden_states, router_logits

File: QEfficient/base/pytorch_transforms.py [hunk 5]
```diff
@@ -107,6 +107,10 @@ def apply(cls, model: nn.Module) -> Tuple[nn.Module, bool]:
             ):
                 for orig_method_name, mapped_method in repl_method_map.items():
```
Comment (quic-rishinr) [inline QEfficient/base/pytorch_transforms.py:108]: can we rename the transform to ExternalModuleMapperTransform? ModuleMethodMapperTransform is misleading

**Review by quic-rishinr — CHANGES_REQUESTED** [review]

File: QEfficient/transformers/models/grok_1/modeling_grok1.py (hunk 2 above)
Comment (quic-rishinr) [reply QEfficient/transformers/models/grok_1/modeling_grok1.py:84]: We are using -10000.0 for all the models now. we need to update it for all the models. Can we update it in a separate PR in one go?

Comment (quic-hemagnih) [reply QEfficient/transformers/models/grok_1/modeling_grok1.py:84]: Yes, but lets atleast define the macro in this PR, why I am suggesting this is because if compiler changes this value in future then we have to make change only at one place. Otherwise we have to make changes at multiple places which is not good practise.

File: QEfficient/base/pytorch_transforms.py (hunk 5 above)
Comment (quic-amitraj) [reply QEfficient/base/pytorch_transforms.py:108]: Done

File: QEfficient/base/pytorch_transforms.py [hunk 6]
```diff
@@ -107,6 +107,10 @@ def apply(cls, model: nn.Module) -> Tuple[nn.Module, bool]:
...
                     setattr(module, orig_method_name, MethodType(mapped_method, module))
+
+                    if hasattr(module, "__qeff_init__"):
```
[the remark this answers is not in the document]
Comment (quic-amitraj) [reply QEfficient/base/pytorch_transforms.py:111]: As we have discusses this is the best way to initialize values to the models modules after loading of the model as wwe are doing in the class ModuleMappingTransform.

**Comment by quic-akuruvil** [comment]
> Add the mode to the validated list and please post the accuracy numbers as well

Also, grok1 2-layer model needs to be added to test files, and please paste with this PR the perplexity numbers in torch and AIC.

File: QEfficient/transformers/models/grok_1/modeling_grok1.py [hunk 7]
```diff
@@ -0,0 +1,356 @@
...
+        elif inputs_embeds is not None:
+            batch_size, seq_length = inputs_embeds.shape[:2]
+        else:
+            raise ValueError("You have to specify either input_ids or inputs_embeds")
+
+        seq_length_with_past = seq_length
+        past_key_values_length = 0
+        if past_key_values is not None:
+            past_key_values_length = past_key_values[0][0].shape[2]
+            seq_length_with_past = seq_length_with_past + past_key_values_length
```
[the remark this answers is not in the document]
Comment (quic-akuruvil) [reply QEfficient/transformers/models/grok_1/modeling_grok1.py:239]: If it is unused, please remove

File: QEfficient/transformers/models/grok_1/modeling_grok1.py (hunk 3 above)
Comment (quic-amitraj) [reply QEfficient/transformers/models/grok_1/modeling_grok1.py:100]: Yes, we can remove this but currently we have this `output_attention` inside all the modeling file. In next transformers version upgrade if everyone will agree will remove from all the modeling file. with that `use_cache` we also have to remove.

File: QEfficient/transformers/models/grok_1/modeling_grok1.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (quic-amitraj) [reply QEfficient/transformers/models/grok_1/modeling_grok1.py:239]: Agree, removed.

File: tests/transformers/models/test_causal_lm_models.py [hunk 8]
```diff
@@ -78,8 +80,10 @@ def load_causal_lm_model(model_config):
         num_hidden_layers=model_config["n_layer"],
         attn_implementation="eager",
         low_cpu_mem_usage=False,
+        trust_remote_code=True,
     )  # Run models for single layers only
     params = sum(p.numel() for p in model_hf.parameters())
+    model_hf.to(torch.float32)
```
Comment (ochougul) [inline tests/transformers/models/test_causal_lm_models.py:86]: is this already added in `modelling_auto.py`

**Review by ochougul — CHANGES_REQUESTED** [review]
Is this tested via infer CLI command?

File: QEfficient/transformers/models/grok_1/modeling_grok1.py [hunk 9]
```diff
@@ -0,0 +1,356 @@
...
+        # Creating experts mask and routing weights masked
+        awesome_experts_mask_1 = (
+            torch.nn.functional.one_hot(selected_experts[:, 0], num_classes=self.num_experts).bool().T.unsqueeze(-1)
+        )
+        awesome_experts_mask_2 = (
+            torch.nn.functional.one_hot(selected_experts[:, 1], num_classes=self.num_experts).bool().T.unsqueeze(-1)
+        )
+
+        gateupout1 = torch.zeros(hidden_states.shape[0], 32768)  # T, hs
+        gateupout2 = torch.zeros(hidden_states.shape[0], 32768)  # T, hs
```
[the remark this answers is not in the document]
Comment (ochougul) [reply QEfficient/transformers/models/grok_1/modeling_grok1.py:120]: Makes sense, save the `ffn_dim` in qeff init assign it in `self` variable and use `self.ffn_dim` here.

You can get `ffn_dim`  from  `self.linear.weight.shape`

**Comment by quic-akuruvil** [comment]
> > > Add the mode to the validated list and please post the accuracy numbers as well
> > 
> > 
> > Also, grok1 2-layer model needs to be added to test files, and please paste with this PR the perplexity numbers in torch and AIC.
> 
> > All
> 
> image
> 
> Output matches end to end on Pytorch, ONNX and AI 100 for single layer.

This is good. But Anuj has suggested to do a first order perplexity analysis from our side for text models, before we close a model, and pass on to accuracy team. Can check with Anuj and run the perplexity scripts if needed.

File: tests/transformers/models/test_causal_lm_models.py (hunk 8 above)
Comment (quic-amitraj) [reply tests/transformers/models/test_causal_lm_models.py:86]: Updated.

**Comment by quic-amitraj** [comment]
> Is this tested via infer CLI command?

It was not supported through infer as this model is not integrated with Transformers. With latest commit added support of this.

**Review by quic-hemagnih — APPROVED** [review]
Looks good to me, we can plan to merge the changes.

File: QEfficient/base/common.py [hunk 10]
```diff
@@ -40,9 +40,13 @@ def from_pretrained(cls, pretrained_model_name_or_path: str, *args, **kwargs) ->
         """
         Downloads HuggingFace model if already doesn't exist locally, returns QEFFAutoModel object based on type of model.
         """
-        config = AutoConfig.from_pretrained(pretrained_model_name_or_path)
+        config = AutoConfig.from_pretrained(pretrained_model_name_or_path, trust_remote_code=True)
```
Comment (quic-rishinr) [inline QEfficient/base/common.py:43]: Lets take this as an argument and enable the trust_remote_code=True. Default lets use trust_remote_code as False.

File: QEfficient/base/modeling_qeff.py [hunk 11]
```diff
@@ -52,7 +52,7 @@ def __init__(self, model: torch.nn.Module) -> None:
         self.onnx_path: Optional[str] = None
         self.qpc_path: Optional[str] = None
         self.qpc_session: Optional[QAICInferenceSession] = None
-
+        model = model.to(torch.float32)
```
Comment (quic-rishinr) [inline QEfficient/base/modeling_qeff.py:55]: please remove this

**Review by quic-rishinr — CHANGES_REQUESTED** [review]

**Review by quic-rishinr — APPROVED** [review]
